### PR TITLE
Added the ability to walk backwards with the continuous hiking pipeline. This is done via the xbox controller

### PR DESCRIPTION
**Pull request by PotatoPeeler3000**
- Added field to message that checks if the user is asking the robot to walk backwards.
- Inside the `ReadyToPlan` state added the check that "if the user wants to go backwards" we set the goal pose to be behind the robot
- Added a parameter for the distance of the goal pose behind the robot
- When walking backwards we want to keep the robot facing the same direction, so we add a couple body path poses so the robot doesn't turn when walking backwards.
- When walking backwards we want to adjust some parameters so we can encourage going backwards, and when not walking backwards we want to change them back
-

### CODE REVIEW

File: ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/perception/RDXContinuousHikingPanel.java [hunk 1]
```diff
@@ -337,13 +337,16 @@ private void publishInputCommandMessage()
       // Setup a bunch of variables to be published in the message
       boolean walkWithKeyboard = ImGui.getIO().getKeyCtrl();
       boolean walkWithController = false;
+      boolean walkBackwards = false;
```
Comment (PotatoPeeler3000) [inline ihmc-high-level-behaviors/src/libgdx/java/us/ihmc/rdx/perception/RDXContinuousHikingPanel.java:340]: Allows for walking backwards

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousHikingStateMachine/ReadyToPlanState.java [hunk 2]
```diff
@@ -155,24 +155,33 @@ public SideDependentList<FramePose3D> getGoalPosesBasedOnPlanningMode()
          {
             if (commandMessage.get().getEnableContinuousHikingWithKeyboard())
             {
-               goalPoses = ContinuousPlannerTools.setRandomizedStraightGoalPoses(continuousPlanner.getWalkingStartMidPose(),
-                                                                                 continuousPlanner.getStartStancePose(),
-                                                                                 (float) continuousHikingParameters.getGoalPoseForwardDistance(),
-                                                                                 (float) continuousHikingParameters.getGoalPoseUpDistance(),
-                                                                                 X_RANDOM_MARGIN,
-                                                                                 NOMINAL_STANCE_WIDTH);
+               goalPoses = ContinuousPlannerTools.setStraightForwardGoalPoses(continuousPlanner.getWalkingStartMidPose(),
```
Comment (PotatoPeeler3000) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousHikingStateMachine/ReadyToPlanState.java:158]: Rename method

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousHikingStateMachine/ReadyToPlanState.java [hunk 3]
```diff
@@ -155,24 +155,33 @@ public SideDependentList<FramePose3D> getGoalPosesBasedOnPlanningMode()
...
+                                                                              (float) continuousHikingParameters.getGoalPoseForwardDistance(),
+                                                                              (float) continuousHikingParameters.getGoalPoseUpDistance(),
+                                                                              X_RANDOM_MARGIN,
+                                                                              NOMINAL_STANCE_WIDTH);
             }
             else if (commandMessage.get().getEnableContinuousHikingWithJoystickController())
             {
+               if (commandMessage.get().getWalkBackwards())
+               {
+                  goalPoses = ContinuousPlannerTools.setStraightBackwardGoalPoses(continuousPlanner.getWalkingStartMidPose(),
```
Comment (PotatoPeeler3000) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousHikingStateMachine/ReadyToPlanState.java:169]: And here is where we check if we want to walk backwards

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousPlanner.java [hunk 4]
```diff
@@ -180,6 +179,22 @@ public FootstepPlan generateAStarFootstepPlan(HeightMapData heightMapData,
       request.setSnapGoalSteps(true);
       request.setAbortIfGoalStepSnappingFails(true);
 
+      // When walking backwards, we want to keep the body facing in the same direction, otherwise the robot will turn as it tries to step backwards
+      if (command.getWalkBackwards())
```
Comment (PotatoPeeler3000) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousPlanner.java:183]: This is where we want to make sure we face the same way when we start walking backwards

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousPlanner.java [hunk 5]
```diff
@@ -180,6 +179,22 @@ public FootstepPlan generateAStarFootstepPlan(HeightMapData heightMapData,
...
+      {
+         FramePose3D bodyMidGoalPose = new FramePose3D();
+         bodyMidGoalPose.interpolate(goalPoses.get(RobotSide.LEFT), goalPoses.get(RobotSide.RIGHT), 0.5);
+         request.getBodyPathWaypoints().add(walkingStartMidPose);
+         request.getBodyPathWaypoints().add(bodyMidGoalPose);
+
+         // To allow walking backwards, we need to change the minimum step length, but this affects other walking, so we only do it when going backwards
+         footstepPlannerParameters.setMinStepLength(-0.3);
```
Comment (PotatoPeeler3000) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousPlanner.java:191]: Specific cases mean we want different parameters

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousPlanner.java [hunk 6]
```diff
@@ -310,7 +325,7 @@ public FootstepPlan generateMonteCarloFootstepPlan(SideDependentList<FramePose3D
       long timeEnd = System.nanoTime();
 
       continuousHikingLogger.appendString(String.format("Total Time: %.3f ms, Plan Size: %d, Visited: %d, Layer Counts: %s",
-                                            (timeEnd - timeStart) / 1e6,
+                                                        (timeEnd - timeStart) / 1e6,
```
Comment (PotatoPeeler3000) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/activeMapping/ContinuousPlanner.java:328]: Formatting

**Review by ihmc-rosie — APPROVED** [review]

**Pull request merged**